File: wallet.go
package main

import (
	"fmt"
	"strconv"
	"strings"
	"os"
)

type Card struct {
	Type           string
	Amount         string
	ExpirationDate string
	CardHolder     string
	CardNumber     string
}

type CIN struct {
	LastName      string
	FirstName     string
	Birthdate     string
	DeliveryDate  string
}

type DriverLicense struct {
	LastName   string
	FirstName  string
	Birthdate  string
	LicenseTypes []string
}

type VisitCard struct {
	LastName    string
	FirstName   string
	Position    string
	CompanyName string
}

var solde = 0.0
var cards []Card
var identityCard []CIN
var driverLicense []DriverLicense
var driverLicenseTypes []string
var visitCards []VisitCard

func showBalance() {
	fmt.Printf("Votre solde actuel: %.2f Ar\n", solde)
}

func depositCash() {
	var amountStr string
	fmt.Print("Veuillez saisir le montant à déposer: Ar ")
	fmt.Scanln(&amountStr)

	amount, err := strconv.ParseFloat(amountStr, 64)
	if err == nil && amount > 0 {
		solde += amount
		fmt.Println("Dépôt effectué!")
	} else {
		fmt.Println("Le montant que vous avez saisi est invalide, veuillez vérifier!")
	}
}

func withdrawCash() {
	var amountStr string
	fmt.Print("Veuillez saisir le montant à retirer: Ar ")
	fmt.Scanln(&amountStr)

	amount, err := strconv.ParseFloat(amountStr, 64)
	if err == nil && amount > 0 {
		if amount <= solde {
			solde -= amount
			fmt.Println("Retrait effectué.")
		} else {
			fmt.Println("Votre solde est insuffisant.")
		}
	} else {
		fmt.Println("Montant invalide.")
	}
}

func depositCard() {
	var cardType, amount, expirationDate, cardHolder, cardNumber string
	fmt.Print("Veuillez saisir le type de carte: ")
	fmt.Scanln(&cardType)
	fmt.Print("Veuillez saisir le montant dans votre carte: ")
	fmt.Scanln(&amount)
	fmt.Print("Veuillez saisir la date d'expiration de la carte: ")
	fmt.Scanln(&expirationDate)
	fmt.Print("Veuillez saisir le propriétaire de cette carte: ")
	fmt.Scanln(&cardHolder)
	fmt.Print("Veuillez saisir le numéro de carte: ")
	fmt.Scanln(&cardNumber)

	card := Card{Type: cardType, Amount: amount, ExpirationDate: expirationDate, CardHolder: cardHolder, CardNumber: cardNumber}
	amountFloat, _ := strconv.ParseFloat(amount, 64)
	solde += amountFloat
	cards = append(cards, card)

	fmt.Println("Votre carte bancaire a bien été déposée.")
}

func recoveredCard() {
	if len(cards) >= 1 {
		fmt.Println("Cartes disponibles à récupérer:")
		for i, card := range cards {
			fmt.Printf("%d- Type: %s, Numéro de carte: %s\n", i+1, card.Type, card.CardNumber)
		}

		var questionStr string
		fmt.Print("Quelle carte voulez-vous récupérer (entrez le numéro correspondant): ")
		fmt.Scanln(&questionStr)

		question, err := strconv.Atoi(questionStr)
		if err == nil && question >= 1 && question <= len(cards) {
			cardIndex := question - 1
			recoveredCard := cards[cardIndex]
			solde -= amountToFloat(recoveredCard.Amount)
			fmt.Printf("La carte récupérée: Type - %s, Numéro de carte - %s\n", recoveredCard.Type, recoveredCard.CardNumber)
			cards = append(cards[:cardIndex], cards[cardIndex+1:]...)
		} else {
			fmt.Println("Veuillez ne saisir que les choix proposés.")
		}
	} else {
		fmt.Println("Vous n'avez aucune carte à récupérer.")
	}
}

func depositCIN() {
	fmt.Println("Veuillez entrer les informations sur votre CIN:")
	var lastName, firstName, birthdate, deliveryDate string
	fmt.Print("Nom: ")
	fmt.Scanln(&lastName)
	fmt.Print("Prenom: ")
	fmt.Scanln(&firstName)
	fmt.Print("Date de naissance: ")
	fmt.Scanln(&birthdate)
	fmt.Print("Date de delivrance: ")
	fmt.Scanln(&deliveryDate)

	identity := CIN{LastName: lastName, FirstName: firstName, Birthdate: birthdate, DeliveryDate: deliveryDate}

	if len(identityCard) == 0 {
		identityCard = append(identityCard, identity)
		fmt.Println("Depot de CIN effectué.")
	} else if len(identityCard) == 1 {
		fmt.Println("Votre CIN devrait être unique.")
	}
}

func retrieveAndShowCIN() {
	if len(identityCard) > 0 {
		fmt.Println("CIN récupérée.")
		fmt.Printf("Les détails de votre CIN:\nNom: %s\nPrenom: %s\nDate de naissance: %s\nDate de delivrance: %s\n",
			identityCard[0].LastName, identityCard[0].FirstName, identityCard[0].Birthdate, identityCard[0].DeliveryDate)
		identityCard = nil
	} else {
		fmt.Println("Aucune carte d'identité identifiée.")
	}
}

func depositDriverLicense() {
	fmt.Println("Veuillez entrez les informations sur votre permis de conduire:")
	var lastName, firstName, birthdate, licenseTypesStr string
	fmt.Print("Nom: ")
	fmt.Scanln(&lastName)
	fmt.Print("Prenom: ")
	fmt.Scanln(&firstName)
	fmt.Print("Date de naissance: ")
	fmt.Scanln(&birthdate)
	fmt.Print("Type de permis (séparez par un virgule si vous avez plus qu'un type | a, b, c): ")
	fmt.Scanln(&licenseTypesStr)

	licenseTypes := strings.Split(licenseTypesStr, ",")

	if len(driverLicense) == 0 {
		driverLicense = append(driverLicense, DriverLicense{LastName: lastName, FirstName: firstName, Birthdate: birthdate, LicenseTypes: licenseTypes})
		fmt.Println("Depot effectué.")
	} else {
		fmt.Println("Votre permis de conduire devrait être unique.")
	}
}

func retrieveAndShowDriverLicense() {
	if len(driverLicense) > 0 {
		fmt.Println("Permis de conduire récupéré.")
		fmt.Printf("Les détails de votre permis:\nNom: %s\nPrenom: %s\nDate de naissance: %s\nType:\n",
			driverLicense[0].LastName, driverLicense[0].FirstName, driverLicense[0].Birthdate)

		for i, licenseType := range driverLicense[0].LicenseTypes {
			fmt.Printf("%d - %s\n", i+1, licenseType)
		}

		driverLicense = nil
	} else {
		fmt.Println("Aucun permis de conduire identifié.")
	}
}

func depositVisitCard() {
	fmt.Println("Veuillez renseigner les informations de votre carte de visite:")
	var lastName, firstName, position, companyName string
	fmt.Print("Nom: ")
	fmt.Scanln(&lastName)
	fmt.Print("Prenom: ")
	fmt.Scanln(&firstName)
	fmt.Print("Poste: ")
	fmt.Scanln(&position)
	fmt.Print("Entreprise: ")
	fmt.Scanln(&companyName)

	if len(visitCards) <= 3 {
		visitCards = append(visitCards, VisitCard{LastName: lastName, FirstName: firstName, Position: position, CompanyName: companyName})
		fmt.Println("Carte de visite enregistrée.")
	} else {
		fmt.Println("Plus d'espace.")
	}
}

func retrieveVisitCards() {
	var cardsToRemove []int

	if len(visitCards) == 0 {
		fmt.Println("Vous n'avez aucune carte de visite dans votre wallet.")
		return
	}

	if len(visitCards) > 1 {
		fmt.Println("Sélectionnez les cartes que vous voulez récupérer (ex: 1,2):")
		for i, visitCard := range visitCards {
			fmt.Printf("%d - %s\n", i+1, visitCard.CompanyName)
		}

		var choice string
		fmt.Print("Carte à récupérer: ")
		fmt.Scanln(&choice)

		if strings.Contains(choice, ",") {
			indices := strings.Split(choice, ",")
			for _, indexStr := range indices {
				index, err := strconv.Atoi(indexStr)
				if err == nil {
					cardsToRemove = append(cardsToRemove, index-1)
				}
			}
		} else {
			index, err := strconv.Atoi(choice)
			if err == nil {
				cardsToRemove = append(cardsToRemove, index-1)
			}
		}
	} else {
		cardsToRemove = append(cardsToRemove, 0)
	}

	for _, index := range cardsToRemove {
		removedCard := visitCards[index]
		fmt.Printf("Carte récupérée: %s\n", removedCard.CompanyName)
		visitCards = append(visitCards[:index], visitCards[index+1:]...)
	}
}

func main() {
	if auth() {
		var choice int

		for {
			fmt.Println(`Bienvenue dans "Wallet", veuillez choisir une action à effectuer:
                1- Voir le solde dans mon compte
                2- Faire un dépôt
                3- Faire un retrait en cash
                4- Deposer une carte bancaire
                5- Recuperer une carte bancaire
                6- Deposer une CIN
                7- Recuperer une CIN
                8- Deposer un permis de conduire
                9- Recuperer et afficher les informations d'un permis de conduire
                10- Deposer une carte de visite
                11- Recuperer une carte de visite
                12- Quitter`)

			fmt.Print("Votre choix: ")
			fmt.Scanln(&choice)

			switch choice {
			case 1:
				showBalance()
			case 2:
				depositCash()
			case 3:
				withdrawCash()
			case 4:
				depositCard()
			case 5:
				recoveredCard()
			case 6:
				depositCIN()
			case 7:
				retrieveAndShowCIN()
			case 8:
				depositDriverLicense()
			case 9:
				retrieveAndShowDriverLicense()
			case 10:
				depositVisitCard()
			case 11:
				retrieveVisitCards()
			case 12:
				fmt.Println("Merci d'utiliser le service Wallet. Au revoir!")
				os.Exit(0)
			default:
				fmt.Println("Choix invalide. Veuillez choisir une option valide.")
			}
		}
	}
}

func auth() bool {
	defaultUsername := "random"
	defaultPassword := "strongpassword"

	var username, userPassword string

	fmt.Print("Veuillez entrer votre nom d'utilisateur: ")
	fmt.Scanln(&username)
	fmt.Print("Veuillez confirmer votre identité en tapant le mot de passe: ")
	fmt.Scanln(&userPassword)

	if username != defaultUsername || userPassword != defaultPassword {
		fmt.Println("Oups, vous n'avez pas accès à ce wallet, désolé.")
		return false
	}
	return true
}

func amountToFloat(amount string) float64 {
	amountFloat, _ := strconv.ParseFloat(amount, 64)
	return amountFloat
}

File: Wallet.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Wallet {
  private static double solde = 0;
  private static List<Card> cards = new ArrayList<>();
  private static List<CIN> identityCard = new ArrayList<>();
  private static List<DriverLicence> driverLicense = new ArrayList<>();
  private static List<String> driverLicenseTypes = new ArrayList<>();
  private static List<VisitCard> visitCards = new ArrayList<>();

  public static void showBalance() {
    System.out.println("Votre solde actuel: " + solde + " Ar");
  }

  public static void depositCash() {
    Scanner scanner = new Scanner(System.in);
    System.out.print("Veuillez saisir le montant à déposer: Ar");
    double amount = scanner.nextDouble();
    if (amount > 0) {
      solde += amount;
      System.out.println("Dépôt effectué!");
    } else {
      System.out.println("Le montant que vous avez saisi est invalide, veuillez vérifier!");
    }
  }

  public static void withdrawCash() {
    Scanner scanner = new Scanner(System.in);
    System.out.print("Veuillez saisir le montant à retirer: Ar");
    double amount = scanner.nextDouble();
    if (amount > 0) {
      if (amount <= solde) {
        solde -= amount;
        System.out.println("Retrait effectué.");
      } else {
        System.out.println("Votre solde est insuffisant.");
      }
    } else {
      System.out.println("Montant invalide.");
    }
  }

  public static void depositCard() {
    Scanner scanner = new Scanner(System.in);
    System.out.print("Veuillez saisir le type de carte: ");
    String type = scanner.nextLine();
    System.out.print("Veuillez saisir le montant dans votre carte: ");
    double amount = scanner.nextDouble();
    scanner.nextLine();
    System.out.print("Veuillez saisir la date d'expiration de la carte: ");
    String expirationDate = scanner.nextLine();
    System.out.print("Veuillez saisir le propriétaire de cette carte: ");
    String cardHolder = scanner.nextLine();
    System.out.print("Veuillez saisir le numéro de carte: ");
    String cardNumber = scanner.nextLine();

    Card card = new Card(type, amount, expirationDate, cardHolder, cardNumber);
    solde += card.getAmount();
    cards.add(card);

    System.out.println("Votre carte bancaire a bien été déposée.");
  }

  public static void recoveredCard() {
    Scanner scanner = new Scanner(System.in);
    if (cards.size() >= 1) {
      System.out.println("Cartes disponibles à récupérer:");
      for (int i = 0; i < cards.size(); i++) {
        System.out.println((i + 1) + "- Type: " + cards.get(i).getType() + ", Numéro de carte: " + cards.get(i).getCardNumber());
      }

      System.out.print("Quelle carte voulez-vous récupérer (entrez le numéro correspondant): ");
      int question = scanner.nextInt();

      if (question >= 1 && question <= cards.size()) {
        int cardIndex = question - 1;
        Card recoveredCard = cards.remove(cardIndex);
        solde -= recoveredCard.getAmount();
        System.out.println("La carte récupérée: Type - " + recoveredCard.getType() + ", Numéro de carte - " + recoveredCard.getCardNumber());
      } else {
        System.out.println("Veuillez ne saisir que les choix proposés.");
      }
    } else {
      System.out.println("Vous n'avez aucune carte à récupérer");
    }
  }

  public static void depositCIN() {
    Scanner scanner = new Scanner(System.in);
    System.out.println("Veuillez entrer les informations sur votre CIN: ");
    System.out.println("Nom: ");
    String lastname = scanner.nextLine();
    System.out.println("Prenom: ");
    String firstname = scanner.nextLine();
    System.out.println("Date de naissance: ");
    String birthdate = scanner.nextLine();
    System.out.println("Date de delivrance: ");
    String deliveryDate = scanner.nextLine();

    CIN identity = new CIN(firstname, lastname, birthdate, deliveryDate);

    if (identityCard.isEmpty()) {
      identityCard.add(identity);
      System.out.println("Depot de CIN effectué.");
    } else if (identityCard.size() == 1) {
      System.out.println("Votre CIN devrait être unique.");
    }
  }

  public static void retrieveAndShowCIN() {
    if (!identityCard.isEmpty()) {
      System.out.println("CIN récupérée.");
      System.out.println("Les détails de votre CIN: ");
      System.out.println("Nom: " + identityCard.get(0).getLastName());
      System.out.println("Prenom: " + identityCard.get(0).getFirstName());
      System.out.println("Date de naissance: " + identityCard.get(0).getBirthdate());
      System.out.println("Date de delivrance: " + identityCard.get(0).getDeliveryDate());

      identityCard.remove(0);
    } else {
      System.out.println("Aucune carte d'identité identifiée.");
    }
  }

  public static void depositDriverLicense() {
    Scanner scanner = new Scanner(System.in);
    System.out.println("Veuillez entrer les informations sur votre permis de conduire: ");
    String lastname = scanner.nextLine();
    String firstname = scanner.nextLine();
    String birthdate = scanner.nextLine();
    String type = scanner.nextLine();

    if (type.contains(",")) {
      String[] types = type.split(",");
      for (String t : types) {
        driverLicenseTypes.add(t.trim());
      }
    }

    if (driverLicense.isEmpty()) {
      driverLicense.add(new DriverLicence(firstname, lastname, birthdate, driverLicenseTypes));
      System.out.println("Depot effectué.");
    } else {
      System.out.println("Votre permis de conduire devrait être unique.");
    }
  }

  public static void retrieveAndShowDriverLicense() {
    if (!driverLicense.isEmpty()) {
      System.out.println("Permis de conduire récupéré.");
      System.out.println("Les détails de votre permis: ");
      System.out.println("Nom: " + driverLicense.get(0).getLastName());
      System.out.println("Prenom: " + driverLicense.get(0).getFirstName());
      System.out.println("Date de naissance: " + driverLicense.get(0).getBirthdate());
      System.out.println("Type: ");

      if (!driverLicenseTypes.isEmpty()) {
        for (int i = 0; i < driverLicenseTypes.size(); i++) {
          System.out.println((i + 1) + " - " + driverLicenseTypes.get(i));
        }
      }
      driverLicense.clear();
    } else {
      System.out.println("Aucun permis de conduire identifié.");
    }
  }

  public static void depositVisitCard() {
    Scanner scanner = new Scanner(System.in);
    System.out.println("Veuillez renseigner les informations de votre carte de visite: ");
    System.out.println("Nom: ");
    String lastname = scanner.nextLine();
    System.out.println("Prenom: ");
    String firstname = scanner.nextLine();
    System.out.println("Fonction: ");
    String fonction = scanner.nextLine();
    System.out.println("Entreprise: ");
    String companyName = scanner.nextLine();

    if (visitCards.size() <= 3) {
      visitCards.add(new VisitCard(lastname, firstname, "",  fonction, companyName));
      System.out.println("Carte de visite enregistrée.");
    } else {
      System.out.println("Plus d'espace.");
    }
  }

  public static void retrieveVisitCards() {
    Scanner scanner = new Scanner(System.in);
    List<Integer> cardsToRemove = new ArrayList<>();

    if (visitCards.isEmpty()) {
      System.out.println("Vous n'avez aucune carte de visite dans votre wallet.");
      return;
    }

    if (visitCards.size() > 1) {
      System.out.println("Sélectionnez les cartes que vous voulez récupérer (ex: 1,2):");
      for (int i = 0; i < visitCards.size(); i++) {
        System.out.println((i + 1) + " - " + visitCards.get(i).getCompanyName());
      }

      System.out.print("Carte à récupérer: ");
      String choice = scanner.nextLine();
      if (choice.contains(",")) {
        String[] indices = choice.split(",");
        for (String index : indices) {
          int cardIndex = Integer.parseInt(index.trim()) - 1;
          if (isValidIndex(cardIndex, visitCards.size())) {
            cardsToRemove.add(cardIndex);
          } else {
            System.out.println("Index invalide: " + (cardIndex + 1));
          }
        }
      } else if (!choice.contains(",")) {
        int cardIndex = Integer.parseInt(choice.trim()) - 1;
        if (isValidIndex(cardIndex, visitCards.size())) {
          cardsToRemove.add(cardIndex);
        } else {
          System.out.println("Index invalide: " + (cardIndex + 1));
        }
      }
    } else {
      cardsToRemove.add(0);
    }

    for (int i = cardsToRemove.size() - 1; i >= 0; i--) {
      int index = cardsToRemove.get(i);
      VisitCard removedCard = visitCards.remove(index);
      System.out.println("Carte récupérée: " + removedCard.getCompanyName());
    }
  }

  private static boolean isValidIndex(int index, int size) {
    return index >= 0 && index < size;
  }
}
